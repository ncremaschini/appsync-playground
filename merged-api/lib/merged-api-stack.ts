import * as cdk from 'aws-cdk-lib';
import * as ssm from 'aws-cdk-lib/aws-ssm';

import { CfnApiKey, CfnDataSource, CfnGraphQLApi, CfnGraphQLSchema, CfnResolver, FieldLogLevel, GraphqlApi, LogConfig } from 'aws-cdk-lib/aws-appsync';

import { Construct } from 'constructs';

export class MergedApiStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);
    
    const servicaAApId = ssm.StringParameter.valueForStringParameter(
      this, '/serviceA/graphQlApiId');     

    const servicaBApId = ssm.StringParameter.valueForStringParameter(
        this, '/serviceB/graphQlApiId');     

    //look up for existing apis to be federated
    const serviceAApis = GraphqlApi.fromGraphqlApiAttributes(this, 'serviceAApis', {
      graphqlApiId: servicaAApId,
    });

    const serviceBApis = GraphqlApi.fromGraphqlApiAttributes(this, 'serviceBApis', {
      graphqlApiId: servicaBApId,
    });
  }
}
