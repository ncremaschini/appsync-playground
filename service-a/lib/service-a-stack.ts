import cdk = require('aws-cdk-lib');

import { AttributeType, BillingMode, StreamViewType, Table, TableEncryption } from 'aws-cdk-lib/aws-dynamodb';
import { CfnApiKey, CfnDataSource, CfnGraphQLApi, CfnGraphQLSchema, CfnResolver } from 'aws-cdk-lib/aws-appsync';
import { ManagedPolicy, Role, ServicePrincipal } from 'aws-cdk-lib/aws-iam';

import { Construct } from 'constructs';

export class ServiceAStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const tableName = "serviceAItems";

    const { itemsTable, itemsTableRole } = this.createTable(tableName);

    const { itemsGraphQLApi, apiKey } = this.createGraphQlApi();

    const apiSchema = this.createGraphQlSchema(itemsGraphQLApi, tableName);    

    const dataSource = this.createGraphQlDatasource(itemsGraphQLApi, itemsTable, itemsTableRole);

    this.createGraphQlResolvers(itemsGraphQLApi, dataSource, tableName, apiSchema);

    this.createStackOutputs(itemsGraphQLApi, apiKey);
  }

  private createGraphQlResolvers(itemsGraphQLApi: cdk.aws_appsync.CfnGraphQLApi, dataSource: cdk.aws_appsync.CfnDataSource, tableName: string, apiSchema: cdk.aws_appsync.CfnGraphQLSchema) {
    const getOneResolver = new CfnResolver(this, "GetOneQueryResolver", {
      apiId: itemsGraphQLApi.attrApiId,
      typeName: "Query",
      fieldName: "getOne",
      dataSourceName: dataSource.name,
      requestMappingTemplate: `{
        "version": "2017-02-28",
        "operation": "GetItem",
        "key": {
          "${tableName}Id": $util.dynamodb.toDynamoDBJson($ctx.args.${tableName}Id)
        }
      }`,
      responseMappingTemplate: `$util.toJson($ctx.result)`,
    });
    getOneResolver.addDependency(apiSchema);
    getOneResolver.addDependency(dataSource);

    const getAllResolver = new CfnResolver(this, "GetAllQueryResolver", {
      apiId: itemsGraphQLApi.attrApiId,
      typeName: "Query",
      fieldName: "all",
      dataSourceName: dataSource.name,
      requestMappingTemplate: `{
        "version": "2017-02-28",
        "operation": "Scan",
        "limit": $util.defaultIfNull($ctx.args.limit, 20),
        "nextToken": $util.toJson($util.defaultIfNullOrEmpty($ctx.args.nextToken, null))
      }`,
      responseMappingTemplate: `$util.toJson($ctx.result)`,
    });
    getAllResolver.addDependency(apiSchema);
    getAllResolver.addDependency(dataSource);

    const saveResolver = new CfnResolver(this, "SaveMutationResolver", {
      apiId: itemsGraphQLApi.attrApiId,
      typeName: "Mutation",
      fieldName: "save",
      dataSourceName: dataSource.name,
      requestMappingTemplate: `{
        "version": "2017-02-28",
        "operation": "PutItem",
        "key": {
          "${tableName}Id": { "S": "$util.autoId()" }
        },
        "attributeValues": {
          "name": $util.dynamodb.toDynamoDBJson($ctx.args.name)
        }
      }`,
      responseMappingTemplate: `$util.toJson($ctx.result)`,
    });
    saveResolver.addDependency(apiSchema);
    saveResolver.addDependency(dataSource);

    const deleteResolver = new CfnResolver(this, "DeleteMutationResolver", {
      apiId: itemsGraphQLApi.attrApiId,
      typeName: "Mutation",
      fieldName: "delete",
      dataSourceName: dataSource.name,
      requestMappingTemplate: `{
        "version": "2017-02-28",
        "operation": "DeleteItem",
        "key": {
          "${tableName}Id": $util.dynamodb.toDynamoDBJson($ctx.args.${tableName}Id)
        }
      }`,
      responseMappingTemplate: `$util.toJson($ctx.result)`,
    });
    deleteResolver.addDependency(apiSchema);
    deleteResolver.addDependency(dataSource);
  }

  private createGraphQlDatasource(itemsGraphQLApi: cdk.aws_appsync.CfnGraphQLApi, itemsTable: cdk.aws_dynamodb.Table, itemsTableRole: cdk.aws_iam.Role) {
    return new CfnDataSource(this, "ServiceAItemsDataSource", {
      apiId: itemsGraphQLApi.attrApiId,
      name: "ServiceAItemsDataSource",
      type: "AMAZON_DYNAMODB",
      dynamoDbConfig: {
        tableName: itemsTable.tableName,
        awsRegion: this.region,
      },
      serviceRoleArn: itemsTableRole.roleArn,
    });
  }

  private createGraphQlSchema(itemsGraphQLApi: cdk.aws_appsync.CfnGraphQLApi, tableName: string) {
    return new CfnGraphQLSchema(this, "serviceAschema", {
      apiId: itemsGraphQLApi.attrApiId,
      definition: `type ${tableName} {
        ${tableName}Id: ID!
        name: String
      }
      type Paginated${tableName} {
        items: [${tableName}!]!
        nextToken: String
      }
      type Query {
        all(limit: Int, nextToken: String): Paginated${tableName}!
        getOne(${tableName}Id: ID!): ${tableName}
      }
      type Mutation {
        save(name: String!): ${tableName}
        delete(${tableName}Id: ID!): ${tableName}
      }
      type Schema {
        query: Query
        mutation: Mutation
      }`,
    });
  }

  private createGraphQlApi() {

    const cloudWatchLogsRole = new Role(this, "ServiceAApiCloudWatchRole", {
      assumedBy: new ServicePrincipal('appsync.amazonaws.com'),
      managedPolicies: [ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSAppSyncPushToCloudWatchLogs')]
    })

    const itemsGraphQLApi = new CfnGraphQLApi(this, "serviceAApi", {
      name: "serviceAApi",
      authenticationType: "API_KEY",
      logConfig: {
        fieldLogLevel: 'ERROR',
        cloudWatchLogsRoleArn: cloudWatchLogsRole.roleArn,
      }
    });

    const apiKey = new CfnApiKey(this, "ServiceAApiKey", {
      apiId: itemsGraphQLApi.attrApiId,
    });
    return { itemsGraphQLApi, apiKey };
  }

  private createStackOutputs(itemsGraphQLApi: cdk.aws_appsync.CfnGraphQLApi, apiKey: cdk.aws_appsync.CfnApiKey) {
    new cdk.CfnOutput(this, 'apiUrl', {
      value: itemsGraphQLApi.attrGraphQlUrl,
      description: 'Graphql invocation url',
      exportName: 'apiUrl',
    });

    new cdk.CfnOutput(this, 'serviceAApiKeyOut', {
      value: apiKey.attrApiKey,
      description: 'api key',
      exportName: 'serviceAApiKey',
    });
  }

  private createTable(tableName: string) {
    const itemsTable = new Table(this, "serviceAItems", {
      tableName: tableName,
      partitionKey: {
        name: `${tableName}Id`,
        type: AttributeType.STRING,
      },
      encryption: TableEncryption.AWS_MANAGED,
      billingMode: BillingMode.PAY_PER_REQUEST,
      stream: StreamViewType.NEW_IMAGE,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });

    const itemsTableRole = new Role(this, "ServiceAItemsDynamoDBRole", {
      assumedBy: new ServicePrincipal("appsync.amazonaws.com"),
    });

    itemsTableRole.addManagedPolicy(
      ManagedPolicy.fromAwsManagedPolicyName("AmazonDynamoDBFullAccess")
    );
    return { itemsTable, itemsTableRole };
  }
}
