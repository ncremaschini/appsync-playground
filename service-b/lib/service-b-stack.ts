import cdk = require('aws-cdk-lib');

import { AttributeType, BillingMode, StreamViewType, Table, TableEncryption } from 'aws-cdk-lib/aws-dynamodb';
import { IResource, LambdaIntegration, MockIntegration, PassthroughBehavior, Period, RestApi } from 'aws-cdk-lib/aws-apigateway';
import { ManagedPolicy, Role, ServicePrincipal } from 'aws-cdk-lib/aws-iam';
import { NodejsFunction, NodejsFunctionProps } from 'aws-cdk-lib/aws-lambda-nodejs';

import { Construct } from 'constructs';
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import { aws_apigateway } from 'aws-cdk-lib';
import { join } from 'path'

export class ServiceBStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const tableName = "serviceBItems";

    const itemsTable = new Table(this, "serviceBItems", {
      tableName: tableName,
      partitionKey: {
        name: `${tableName}Id`,
        type: AttributeType.STRING,
      },
      encryption: TableEncryption.AWS_MANAGED,
      billingMode: BillingMode.PAY_PER_REQUEST,
      stream: StreamViewType.NEW_IMAGE,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });

    const itemsTableRole = new Role(this, "serviceBItemsDynamoDBRole", {
      assumedBy: new ServicePrincipal("appsync.amazonaws.com"),
    });

    itemsTableRole.addManagedPolicy(
      ManagedPolicy.fromAwsManagedPolicyName("AmazonDynamoDBFullAccess")
    );

    const nodeJsFunctionProps: NodejsFunctionProps = {
      bundling: {
        externalModules: [
          'aws-sdk', // Use the 'aws-sdk' available in the Lambda runtime
        ],
      },
      depsLockFilePath: join(__dirname, 'lambdas', 'package-lock.json'),
      environment: {
        PRIMARY_KEY: 'itemId',
        TABLE_NAME: itemsTable.tableName,
      },
      runtime: Runtime.NODEJS_18_X,
    }

    console.log(__dirname);
    
    // Create a Lambda function for each of the CRUD operations
    const getOneLambda = new NodejsFunction(this, 'getOneItemFunction', {
      entry: join(__dirname, 'lambdas', 'get-one.ts'),
      ...nodeJsFunctionProps,
    });
    const getAllLambda = new NodejsFunction(this, 'getAllItemsFunction', {
      entry: join(__dirname, 'lambdas', 'get-all.ts'),
      ...nodeJsFunctionProps,
    });
    const createOneLambda = new NodejsFunction(this, 'createItemFunction', {
      entry: join(__dirname, 'lambdas', 'create.ts'),
      ...nodeJsFunctionProps,
    });
    const updateOneLambda = new NodejsFunction(this, 'updateItemFunction', {
      entry: join(__dirname, 'lambdas', 'update-one.ts'),
      ...nodeJsFunctionProps,
    });
    const deleteOneLambda = new NodejsFunction(this, 'deleteItemFunction', {
      entry: join(__dirname, 'lambdas', 'delete-one.ts'),
      ...nodeJsFunctionProps,
    });

    // Grant the Lambda function read access to the DynamoDB table
    itemsTable.grantReadWriteData(getAllLambda);
    itemsTable.grantReadWriteData(getOneLambda);
    itemsTable.grantReadWriteData(createOneLambda);
    itemsTable.grantReadWriteData(updateOneLambda);
    itemsTable.grantReadWriteData(deleteOneLambda);

    // Integrate the Lambda functions with the API Gateway resource
    const getAllIntegration = new LambdaIntegration(getAllLambda);
    const createOneIntegration = new LambdaIntegration(createOneLambda);
    const getOneIntegration = new LambdaIntegration(getOneLambda);
    const updateOneIntegration = new LambdaIntegration(updateOneLambda);
    const deleteOneIntegration = new LambdaIntegration(deleteOneLambda);


    // Create an API Gateway resource for each of the CRUD operations
    const api = new RestApi(this, 'serviceBApi', {
      restApiName: 'serviceBApi'
    });

    const apiKeyName = "dev-key"

    const apiKey = new aws_apigateway.ApiKey(this, `APIkey`, {
      apiKeyName,
      description: `APIKey used by my api to do awesome stuff`,
      enabled: true,
    })

    new cdk.CfnOutput(this, 'apiKey', {
      value: apiKey.keyId,
      description: 'api key',
      exportName: 'apiKey',
    });

    const usagePlanProps: aws_apigateway.UsagePlanProps = {
      name: "devUsagePlan",
      apiStages: [{api: api, stage: api.deploymentStage}],
      throttle: {burstLimit: 500, rateLimit: 1000}, quota: {limit: 10000000, period: Period.MONTH}
    }

    api.addUsagePlan(`devUsagePlan`, usagePlanProps).addApiKey(apiKey);

    new cdk.CfnOutput(this, 'apiUrl', {
      value: api.url,
      description: 'rest api invocation url',
      exportName: 'apiUrl',
    });

    const items = api.root.addResource('items');
    items.addMethod('GET', getAllIntegration);
    items.addMethod('POST', createOneIntegration);
    addCorsOptions(items);

    const singleItem = items.addResource('{id}');
    singleItem.addMethod('GET', getOneIntegration);
    singleItem.addMethod('PATCH', updateOneIntegration);
    singleItem.addMethod('DELETE', deleteOneIntegration);
    addCorsOptions(singleItem);
  }
}

export function addCorsOptions(apiResource: IResource) {
  apiResource.addMethod('OPTIONS', new MockIntegration({
    // In case you want to use binary media types, uncomment the following line
    // contentHandling: ContentHandling.CONVERT_TO_TEXT,
    integrationResponses: [{
      statusCode: '200',
      responseParameters: {
        'method.response.header.Access-Control-Allow-Headers': "'Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token,X-Amz-User-Agent'",
        'method.response.header.Access-Control-Allow-Origin': "'*'",
        'method.response.header.Access-Control-Allow-Credentials': "'false'",
        'method.response.header.Access-Control-Allow-Methods': "'OPTIONS,GET,PUT,POST,DELETE'",
      },
    }],
    // In case you want to use binary media types, comment out the following line
    passthroughBehavior: PassthroughBehavior.NEVER,
    requestTemplates: {
      "application/json": "{\"statusCode\": 200}"
    },
  }), {
    methodResponses: [{
      statusCode: '200',
      responseParameters: {
        'method.response.header.Access-Control-Allow-Headers': true,
        'method.response.header.Access-Control-Allow-Methods': true,
        'method.response.header.Access-Control-Allow-Credentials': true,
        'method.response.header.Access-Control-Allow-Origin': true,
      },
    }]
  })
}

