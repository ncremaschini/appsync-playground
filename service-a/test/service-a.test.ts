import * as ServiceA from '../lib/service-a-stack';
import * as cdk from 'aws-cdk-lib';

import { AttributeType } from 'aws-cdk-lib/aws-dynamodb';
import { Template } from 'aws-cdk-lib/assertions';

test("All resources created", () => {
  const app = new cdk.App();
  // WHEN
  const stack = new ServiceA.ServiceAStack(app, "TestStack");
  // THEN
  const template = Template.fromStack(stack);

  expect(template).toMatchSnapshot();

  template.hasResourceProperties('AWS::DynamoDB::Table', {
    TableName: 'serviceAItems',
    KeySchema: [
        {
            "AttributeName": "serviceAItemsId",
            "KeyType": "HASH",
        }],
    SSESpecification: {
      SSEEnabled: true
    }
  });
  template.resourceCountIs('AWS::DynamoDB::Table', 1);
});
