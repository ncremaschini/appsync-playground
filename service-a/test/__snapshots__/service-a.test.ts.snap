// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`All resources created 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "DeleteMutationResolver": {
      "DependsOn": [
        "serviceAschema",
      ],
      "Properties": {
        "ApiId": {
          "Fn::GetAtt": [
            "serviceAApi",
            "ApiId",
          ],
        },
        "DataSourceName": "ServiceAItemsDataSource",
        "FieldName": "delete",
        "RequestMappingTemplate": "{
        "version": "2017-02-28",
        "operation": "DeleteItem",
        "key": {
          "serviceAItemsId": $util.dynamodb.toDynamoDBJson($ctx.args.serviceAItemsId)
        }
      }",
        "ResponseMappingTemplate": "$util.toJson($ctx.result)",
        "TypeName": "Mutation",
      },
      "Type": "AWS::AppSync::Resolver",
    },
    "GetAllQueryResolver": {
      "DependsOn": [
        "serviceAschema",
      ],
      "Properties": {
        "ApiId": {
          "Fn::GetAtt": [
            "serviceAApi",
            "ApiId",
          ],
        },
        "DataSourceName": "ServiceAItemsDataSource",
        "FieldName": "all",
        "RequestMappingTemplate": "{
        "version": "2017-02-28",
        "operation": "Scan",
        "limit": $util.defaultIfNull($ctx.args.limit, 20),
        "nextToken": $util.toJson($util.defaultIfNullOrEmpty($ctx.args.nextToken, null))
      }",
        "ResponseMappingTemplate": "$util.toJson($ctx.result)",
        "TypeName": "Query",
      },
      "Type": "AWS::AppSync::Resolver",
    },
    "GetOneQueryResolver": {
      "DependsOn": [
        "serviceAschema",
      ],
      "Properties": {
        "ApiId": {
          "Fn::GetAtt": [
            "serviceAApi",
            "ApiId",
          ],
        },
        "DataSourceName": "ServiceAItemsDataSource",
        "FieldName": "getOne",
        "RequestMappingTemplate": "{
        "version": "2017-02-28",
        "operation": "GetItem",
        "key": {
          "serviceAItemsId": $util.dynamodb.toDynamoDBJson($ctx.args.serviceAItemsId)
        }
      }",
        "ResponseMappingTemplate": "$util.toJson($ctx.result)",
        "TypeName": "Query",
      },
      "Type": "AWS::AppSync::Resolver",
    },
    "SaveMutationResolver": {
      "DependsOn": [
        "serviceAschema",
      ],
      "Properties": {
        "ApiId": {
          "Fn::GetAtt": [
            "serviceAApi",
            "ApiId",
          ],
        },
        "DataSourceName": "ServiceAItemsDataSource",
        "FieldName": "save",
        "RequestMappingTemplate": "{
        "version": "2017-02-28",
        "operation": "PutItem",
        "key": {
          "serviceAItemsId": { "S": "$util.autoId()" }
        },
        "attributeValues": {
          "name": $util.dynamodb.toDynamoDBJson($ctx.args.name)
        }
      }",
        "ResponseMappingTemplate": "$util.toJson($ctx.result)",
        "TypeName": "Mutation",
      },
      "Type": "AWS::AppSync::Resolver",
    },
    "ServiceAItemsDataSource": {
      "Properties": {
        "ApiId": {
          "Fn::GetAtt": [
            "serviceAApi",
            "ApiId",
          ],
        },
        "DynamoDBConfig": {
          "AwsRegion": {
            "Ref": "AWS::Region",
          },
          "TableName": {
            "Ref": "serviceAItems03DA76B3",
          },
        },
        "Name": "ServiceAItemsDataSource",
        "ServiceRoleArn": {
          "Fn::GetAtt": [
            "ServiceAItemsDynamoDBRole196791C2",
            "Arn",
          ],
        },
        "Type": "AMAZON_DYNAMODB",
      },
      "Type": "AWS::AppSync::DataSource",
    },
    "ServiceAItemsDynamoDBRole196791C2": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "appsync.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AmazonDynamoDBFullAccess",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "serviceAApi": {
      "Properties": {
        "AuthenticationType": "API_KEY",
        "Name": "serviceAApi",
      },
      "Type": "AWS::AppSync::GraphQLApi",
    },
    "serviceAApiKey": {
      "Properties": {
        "ApiId": {
          "Fn::GetAtt": [
            "serviceAApi",
            "ApiId",
          ],
        },
      },
      "Type": "AWS::AppSync::ApiKey",
    },
    "serviceAItems03DA76B3": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "serviceAItemsId",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "serviceAItemsId",
            "KeyType": "HASH",
          },
        ],
        "SSESpecification": {
          "SSEEnabled": true,
        },
        "StreamSpecification": {
          "StreamViewType": "NEW_IMAGE",
        },
        "TableName": "serviceAItems",
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
    "serviceAschema": {
      "Properties": {
        "ApiId": {
          "Fn::GetAtt": [
            "serviceAApi",
            "ApiId",
          ],
        },
        "Definition": "type serviceAItems {
        serviceAItemsId: ID!
        name: String
      }
      type PaginatedserviceAItems {
        items: [serviceAItems!]!
        nextToken: String
      }
      type Query {
        all(limit: Int, nextToken: String): PaginatedserviceAItems!
        getOne(serviceAItemsId: ID!): serviceAItems
      }
      type Mutation {
        save(name: String!): serviceAItems
        delete(serviceAItemsId: ID!): serviceAItems
      }
      type Schema {
        query: Query
        mutation: Mutation
      }",
      },
      "Type": "AWS::AppSync::GraphQLSchema",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
